cpp/src/arrow/compute/kernels/vector_run_end_encode_test.cc

```diff
@@ -0,0 +1,211 @@
+  ASSERT_ARRAYS_EQUAL(*ArrayFromJSON(run_ends_type, data.expected_run_ends_json),
+                      *run_ends_array);
+  ASSERT_ARRAYS_EQUAL(*values_array, *data.expected_values);
+  ASSERT_EQ(encoded->buffers.size(), 1);
+  ASSERT_EQ(encoded->buffers[0], NULLPTR);
+  ASSERT_EQ(encoded->child_data.size(), 2);
+  ASSERT_EQ(run_ends_array->data()->buffers[0], NULLPTR);
+  ASSERT_EQ(run_ends_array->length(), data.expected_values->length());
+  ASSERT_EQ(run_ends_array->offset(), 0);
+  ASSERT_EQ(encoded->length, data.input->length());
+  ASSERT_EQ(encoded->offset, 0);
+  ASSERT_EQ(*encoded->type, RunEndEncodedType(run_ends_type, data.input->type()));
+  ASSERT_EQ(encoded->null_count, 0);
+
+  ASSERT_OK_AND_ASSIGN(Datum decoded_datum, RunEndDecode(encoded));
+  auto decoded = decoded_datum.make_array();
+  ASSERT_OK(decoded->ValidateFull());
+  ASSERT_ARRAYS_EQUAL(*decoded, *data.input);
+}
+
+// Encoding an input with an offset results in a completely new encoded array without an
+// offset. This means The EncodeDecodeArray test will never actually decode an array
+// with an offset, even though we have inputs with offsets. This test slices one element
+// off the encoded array and decodes that.
+TEST_P(TestRunEndEncodeDecode, DecodeWithOffset) {
+  auto [data, run_ends_type] = GetParam();
+  if (data.input->length() == 0) {
+    return;
+  }
+
+  ASSERT_OK_AND_ASSIGN(Datum encoded_datum,
+                       RunEndEncode(data.input, RunEndEncodeOptions(run_ends_type)));
+
+  auto encoded = encoded_datum.array();
+  ASSERT_OK_AND_ASSIGN(Datum datum_without_first,
+                       RunEndDecode(encoded->Slice(1, encoded->length - 1)));
+  ASSERT_OK_AND_ASSIGN(Datum datum_without_last,
+                       RunEndDecode(encoded->Slice(0, encoded->length - 1)));
+  auto array_without_first = datum_without_first.make_array();
+  auto array_without_last = datum_without_last.make_array();
+  ASSERT_OK(array_without_first->ValidateFull());
+  ASSERT_OK(array_without_last->ValidateFull());
+  ASSERT_ARRAYS_EQUAL(*array_without_first, *data.input->Slice(1));
+  ASSERT_ARRAYS_EQUAL(*array_without_last,
+                      *data.input->Slice(0, data.input->length() - 1));
+}
+
+// This test creates an run-end encoded array with an offset in the child array, which
+// removes the first run in the test data.
+TEST_P(TestRunEndEncodeDecode, DecodeWithOffsetInChildArray) {
+  auto [data, run_ends_type] = GetParam();
+
+  ASSERT_OK_AND_ASSIGN(Datum encoded_datum,
+                       RunEndEncode(data.input, RunEndEncodeOptions(run_ends_type)));
+
+  auto encoded = encoded_datum.array();
+  this->AddArtificialOffsetInChildArray(encoded.get(), 100);
+  ASSERT_OK_AND_ASSIGN(Datum datum_without_first, RunEndDecode(encoded));
+  auto array_without_first = datum_without_first.make_array();
+  ASSERT_OK(array_without_first->ValidateFull());
+  ASSERT_ARRAYS_EQUAL(*array_without_first, *data.input);
+}
+
+INSTANTIATE_TEST_SUITE_P(
```

Comment:
Is it possible to test scalars and chunked arrays here, too?

Reply:
The implementation doesn't handle `RunEndEncodedScalar`, so I will have to add that first ;-) Thanks for reminding me of that.

Reply:
Ah, whoops!

Reply:
Well, vector functions don't really deal with scalars, so scalars are not needed. cc @westonpace

I'm looking into chunked arrays now.

Reply:
Done. Lots of chunked array tests!